packages/qwik-city/src/middleware/request-handler/user-response.ts

```diff
@@ -50,62 +52,94 @@ export function runQwikCity<T>(
+  rebuildRouteInfo: RebuildRouteInfoInternal,
+  resolve: (value: any) => void
+) {
+  let rewriteAttempt = 1;
+  let didRewrite = false;
+
+  async function _runNext() {
+    didRewrite = false;
+
+    try {
+      // Run all middlewares
+      await requestEv.next();
+    } catch (e) {
+      if (e instanceof RedirectMessage) {
+        const stream = requestEv.getWritableStream();
+        await stream.close();
+      } else if (e instanceof RewriteMessage) {
+        if (rewriteAttempt > 5) {
+          throw new Error(`Rewrite failed - Max rewrite attempts reached: ${rewriteAttempt - 1}`);
         }
-      }
-    } else if (!(e instanceof AbortMessage)) {
-      if (getRequestMode(requestEv) !== 'dev') {
-        try {
-          if (!requestEv.headersSent) {
-            requestEv.headers.set('content-type', 'text/html; charset=utf-8');
-            requestEv.cacheControl({ noCache: true });
-            requestEv.status(500);
+
+        didRewrite = true;
+      } else if (e instanceof ServerError) {
+        if (!requestEv.headersSent) {
+          const status = e.status as StatusCodes;
+          const accept = requestEv.request.headers.get('Accept');
+          if (accept && !accept.includes('text/html')) {
+            const qwikSerializer = requestEv[RequestEvQwikSerializer];
+            requestEv.headers.set('Content-Type', 'application/qwik-json');
+            requestEv.send(status, await qwikSerializer._serializeData(e.data, true));
+          } else {
+            const html = getErrorHtml(e.status, e.data);
+            requestEv.html(status, html);
           }
-          const stream = requestEv.getWritableStream();
-          if (!stream.locked) {
-            const writer = stream.getWriter();
-            await writer.write(encoder.encode(minimalHtmlResponse(500, 'Internal Server Error')));
-            await writer.close();
+        }
+      } else if (!(e instanceof AbortMessage)) {
+        if (getRequestMode(requestEv) !== 'dev') {
+          try {
+            if (!requestEv.headersSent) {
+              requestEv.headers.set('content-type', 'text/html; charset=utf-8');
+              requestEv.cacheControl({ noCache: true });
+              requestEv.status(500);
+            }
+            const stream = requestEv.getWritableStream();
+            if (!stream.locked) {
+              const writer = stream.getWriter();
+              await writer.write(encoder.encode(minimalHtmlResponse(500, 'Internal Server Error')));
+              await writer.close();
+            }
+          } catch {
+            console.error('Unable to render error page');
           }
-        } catch {
-          console.error('Unable to render error page');
         }
+        return e;
       }
-      return e;
     }
+
+    return undefined;
+  }
+
+  try {
+    let runResult = await _runNext();
+    if (didRewrite) {
+      rewriteAttempt += 1;
+      const url = new URL(requestEv.headers.get('Rewrite-Location')!);
```

Comment:
I'm confused, why use the headers as a side-channel? Why not store it on the error object?